x-pack/lib/config_management/elasticsearch_source.rb

```diff
@@ -53,7 +53,17 @@ def config_conflict?
 
       # decide using system indices api (7.10+) or legacy api (< 7.10) base on elasticsearch server version
       def get_pipeline_fetcher
-        response = client.get("/")
+        response = nil
+        loop do
+          begin
+            response = client.get("/")
+            break
+          rescue ::LogStash::Outputs::ElasticSearch::HttpClient::Pool::HostUnreachableError => e
+            # continue looping until the client connects to a good host or NoConnectionAvailableError is raised.
+            failed_host = e.url
+            logger.info("Reiterate with next host, client should remove from the pool ", {:host => failed_host})
+          end
+        end
```

Comment:
We could use `Stud::try`, which is already available and supplies us with exponential backoff.

```suggestion
        response = Stud::try(loop, ::LogStash::Outputs::ElasticSearch::HttpClient::Pool::HostUnreachableError) { client.get("/") }
```

If we want a better handler, we can subclass `Stud::Try`:

~~~ ruby
class ElasticsearchRequestHandler < Stud::Try
  def initialize(logger, name)
    @logger = logger
    @name = name
  end

  def log_failure(exception, fail_count, message)
    @logger.warn("Attempt to #{@name} failed. #{message}", fail_count: fail_count, exception: e.message)
  end
end
~~~

Which would make the above:

```suggestion
        retry_handler = ElasticsearchRequestHandler.new(logger, 'Elasticsearch capability detection')
        response = retry_handler.try(loop, ::LogStash::Outputs::ElasticSearch::HttpClient::Pool::HostUnreachableError) { client.get("/") }
```

Reply:
Thanks @yaauie, nice suggestion